File: go/src/460.lfu缓存.go
/*
 * @lc app=leetcode.cn id=460 lang=golang
 *
 * [460] LFU缓存
 */

// @lc code=start

// 双哈希表
// 一个以frequency为key，一个以数据的key为key
type node struct {
	key int
	value int
	freq int
}

type LFUCache struct {
	capacity int
	count int
	minFreq int
	freq map[int]*list.List
	hash map[int]*list.Element
}


func Constructor(capacity int) LFUCache {
	return LFUCache{
		capacity: capacity,
		freq: make(map[int]*list.List),
		hash: make(map[int]*list.Element),
		minFreq: 1,
	}
}


func (this *LFUCache) Get(key int) int {
	if ele, ok := this.hash[key]; ok {
		n := ele.Value.(*node)
		//fmt.Println("get before delete: ",this.freq[n.freq].Len())
		//this.freq[n.freq].Remove(ele)
		//fmt.Println("after delete: ",this.freq[n.freq].Len())
		for i := this.freq[n.freq].Front(); i != nil; i = i.Next() {
			n := i.Value.(*node)
			if key == n.key {
				this.freq[n.freq].Remove(i)
				break
			}
		}
		
		if this.minFreq == n.freq && this.freq[n.freq].Len() == 0 {
			this.minFreq++
		}

		n.freq++
		if l, ok := this.freq[n.freq]; ok {
			l.PushFront(n)
		} else {
			this.freq[n.freq] = list.New()
			this.freq[n.freq].PushFront(n)
		}

		return n.value
	}
	return -1
}


func (this *LFUCache) Put(key int, value int)  {
	if ele, ok := this.hash[key]; ok {
		n := ele.Value.(*node)
		//fmt.Println("put before delete: ",this.freq[n.freq].Len())
		//this.freq[n.freq].Remove(ele) // BUG！！！ 有可能删不掉
		//fmt.Println("after delete: ",this.freq[n.freq].Len())
		for i := this.freq[n.freq].Front(); i != nil; i = i.Next() {
			n := i.Value.(*node)
			if key == n.key {
				this.freq[n.freq].Remove(i)
				break
			}
		}

		if this.minFreq == n.freq && this.freq[n.freq].Len() == 0 {
			this.minFreq++
		}

		n.freq++
		n.value = value
		if l, ok := this.freq[n.freq]; ok {
			l.PushFront(n)
		} else {
			this.freq[n.freq] = list.New()
			this.freq[n.freq].PushFront(n)
		}
	} else {
		if this.capacity == 0 {
			return
		}

		if this.count == this.capacity {
			element := this.freq[this.minFreq].Back()
			n := element.Value.(*node)
			delete(this.hash, n.key)
			this.freq[this.minFreq].Remove(element)
			
			this.count--
		}

		this.count++
		n := &node{
			key: key,
			value: value,
			freq: 1,
		}

		if l, ok := this.freq[1]; ok {
			l.PushFront(n)
		} else {
			this.freq[1] = list.New()
			this.freq[1].PushFront(n)
		}

		this.hash[key] = this.freq[1].Front()
		this.minFreq = 1
	}
}

type node struct {
	key int
	val int
	count int
}

// 哈希表加双向链表
type LFUCache1 struct {
	l *list.List
	hash map[int]*list.Element
	count int
	capacity int
}


func Constructor1(capacity int) LFUCache1 {
	return LFUCache1{
		capacity: capacity,
		hash: make(map[int]*list.Element),
		l: list.New(),
	}
}


func (this *LFUCache1) Get(key int) int {
	if element, ok := this.hash[key]; ok {
		n := element.Value.(*node)
		n.count++
		for i := element.Next(); i != nil; i = i.Next() {
			n2 := i.Value.(*node)
			if n2.count > n.count {
				this.l.MoveBefore(element, i)
				return n.val
			}
		}
		this.l.MoveToBack(element)
		return n.val
	}
	return -1
}


func (this *LFUCache1) Put(key int, value int)  {
	if ele, ok := this.hash[key]; ok {
		n := ele.Value.(*node)
		n.val = value
		n.count++
		for i := ele.Next(); i != nil; i = i.Next() {
			n2 := i.Value.(*node)
			if n2.count > n.count {
				this.l.MoveBefore(ele, i)
				return
			}
		}
		this.l.MoveToBack(ele)
	} else {
		if this.count >= this.capacity {
			if this.count == 0 {
				return
			}
			
			this.count--
			head := this.l.Front()
			n := head.Value.(*node)
			delete(this.hash, n.key)
			this.l.Remove(head)
		}

		this.count++
		n := &node{
			key: key,
			val: value,
			count: 1,
		}

		for i := this.l.Front(); i != nil; i = i.Next() {
			n2 := i.Value.(*node)
			if n2.count > n.count {
				this.l.InsertBefore(n, i)
				this.hash[key] = i.Prev()
				return
			}
		}
		
		this.l.PushBack(n)
		this.hash[key] = this.l.Back()
	}
}


/**
 * Your LFUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */
// @lc code=end

File: rust/src/460.lfu-缓存.rs
/*
 * @lc app=leetcode.cn id=460 lang=rust
 *
 * [460] LFU 缓存
 */

// @lc code=start
use std::cell::{Ref, RefCell};
use std::collections::{BinaryHeap, HashMap, HashSet};
use std::rc::Rc;

struct Node {
    key: i32,
    val: i32,
    freq: i32,
    prev: Option<Rc<RefCell<Node>>>,
    next: Option<Rc<RefCell<Node>>>,
}

impl Node {
    pub fn new(key: i32, value: i32) -> Rc<RefCell<Node>> {
        Rc::new(RefCell::new(Self {
            key: key,
            val: value,
            freq: 1,
            prev: None,
            next: None,
        }))
    }
}

struct LFUCache {
    capacity: usize,
    min_freq: i32,
    // key to node
    map: HashMap<i32, Rc<RefCell<Node>>>,
    // freq to node list
    freq: HashMap<i32, Rc<RefCell<Node>>>,
}

/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl LFUCache {
    pub fn new(capacity: i32) -> Self {
        Self {
            capacity: capacity as usize,
            min_freq: 0,
            map: HashMap::new(),
            freq: HashMap::new(),
        }
    }

    pub fn get(&mut self, key: i32) -> i32 {
        if let Some(node) = self.get_node(key) {
            return node.borrow().val;
        }

        -1
    }

    pub fn put(&mut self, key: i32, value: i32) {
        if let Some(node) = self.get_node(key) {
            node.borrow_mut().val = value;
            return;
        }

        if self.map.len() == self.capacity {
            let dummy = self.freq.get(&self.min_freq).unwrap();
            let dummy2 = &dummy.clone();
            let last_node = dummy.borrow().prev.clone().unwrap();
            let key = last_node.borrow().key;
            self.map.remove(&key);
            self.remove(last_node);
            if Rc::ptr_eq(dummy2, dummy2.borrow().prev.as_ref().unwrap()) {
                self.freq.remove(&self.min_freq);
            }
        }
        let node = Node::new(key, value);
        self.map.insert(key, node.clone());
        self.push_front(1, node.clone());
        self.min_freq = 1;
    }

    fn get_node(&mut self, key: i32) -> Option<Rc<RefCell<Node>>> {
        if let Some(node) = self.map.get(&key) {
            let node = node.clone();
            self.remove(node.clone());

            let freq = node.borrow().freq;
            let dummy = self.freq.get(&freq).unwrap();
            if Rc::ptr_eq(dummy, dummy.borrow().prev.as_ref().unwrap()) {
                self.freq.remove(&freq);
                if self.min_freq == freq {
                    self.min_freq += 1;
                }
            }

            node.borrow_mut().freq += 1;
            self.push_front(freq + 1, node.clone());
            return Some(node);
        }

        None
    }

    fn remove(&mut self, node: Rc<RefCell<Node>>) {
        let prev = node.borrow().prev.clone().unwrap();
        let next = node.borrow().next.clone().unwrap();
        prev.borrow_mut().next = Some(next.clone());
        next.borrow_mut().prev = Some(prev.clone());
    }

    fn push_front(&mut self, freq: i32, node: Rc<RefCell<Node>>) {
        let list = self.freq.entry(freq).or_insert_with(|| Self::new_list());
        let next = list.borrow().next.clone();
        node.borrow_mut().prev = Some(list.clone());
        node.borrow_mut().next = next.clone();
        list.borrow_mut().next = Some(node.clone());
        next.unwrap().borrow_mut().prev = Some(node);
    }

    fn new_list() -> Rc<RefCell<Node>> {
        let dummy = Node::new(0, 0);
        dummy.borrow_mut().prev = Some(dummy.clone());
        dummy.borrow_mut().next = Some(dummy.clone());
        dummy
    }
}

/**
 * Your LFUCache object will be instantiated and called as such:
 * let obj = LFUCache::new(capacity);
 * let ret_1: i32 = obj.get(key);
 * obj.put(key, value);
 */
// @lc code=end

